### PR TITLE
EIP-211:implement OPCODE RETURNDATASIZE and RETURNDATACOPY

### PR DESCRIPTION
**Pull request by gasolin**
Implement EIP-211 https://github.com/ethereum/EIPs/pull/211
failed test cases from 483 -> 481

### CODE REVIEW

File: lib/opFns.js [hunk 1]
```diff
@@ -318,6 +318,19 @@ module.exports = {
...
+  },
+  RETURNDATACOPY: function (memOffset, codeOffset, length, runState) {
+    memOffset = utils.bufferToInt(memOffset)
+    codeOffset = utils.bufferToInt(codeOffset)
+    length = utils.bufferToInt(length)
+
+    if (codeOffset + length > runState.lastReturned.length) {
+      trap(ERROR.RETURNDATACOPY_OUT_OF_RANGE)
+    }
+    memStore(runState, memOffset, utils.toBuffer(runState.last_returned), codeOffset, length, false)
```
Comment (jwasinger) [inline lib/opFns.js]: this should be `lastReturned` instead of `last_returned`.

Comment (jwasinger) [inline lib/opFns.js]: Change `last_returned` to `lastReturned`

File: lib/opFns.js [hunk 2]
```diff
@@ -318,6 +318,19 @@ module.exports = {
...
   },
+  RETURNDATASIZE: function (runState) {
+    return utils.intToBuffer(runState.lastReturned.length)
+  },
+  RETURNDATACOPY: function (memOffset, codeOffset, length, runState) {
+    memOffset = utils.bufferToInt(memOffset)
+    codeOffset = utils.bufferToInt(codeOffset)
+    length = utils.bufferToInt(length)
+
+    if (codeOffset + length > runState.lastReturned.length) {
```
Comment (jwasinger) [inline lib/opFns.js]: Change L329-L333 to 
```
    memStore(runState, memOffset, utils.toBuffer(runState.lastReturned), codeOffset, length, false)
    subGas(runState, new BN(fees.copyGas.v * Math.ceil(length / 32)))
```

File: lib/opFns.js [hunk 3]
```diff
@@ -908,8 +921,12 @@ function makeCall (runState, callOptions, localOpts, cb) {
       for (var i = 0; i < Math.min(localOpts.outLength, results.vm.return.length); i++) {
         runState.memory[localOpts.outOffset + i] = results.vm.return[i]
       }
+
+      if (['CALL', 'CALLCODE', 'DELEGATECALL'].indexOf(runState.opName) >= -1) {
```
Comment (jwasinger) [inline lib/opFns.js]: L925 should be rewritten as:
```
      switch(runState.opName) {
        case 'CALL':
        case 'CALLCODE':
        case 'DELEGATECALL':
          runState.lastReturned = results.vm.return
      }   
```

Easier to understand.  Probably faster as well.

**Review by jwasinger — DISMISSED** [review]
test

File: lib/opFns.js (hunk 2 above)
Comment (jwasinger) [inline lib/opFns.js]: The end of this function (L329 onwards) should be:
```
memStore(runState, memOffset, utils.toBuffer(runState.lastReturned), codeOffset, length, false)
subGas(runState, new BN(fees.copyGas.v * Math.ceil(length / 32)))
```

**Comment by gasolin** [comment]
updated, now failed test cases reduced from 483 -> 481

**Comment by gasolin** [comment]
updated, now failed test cases reduced from 483 -> 479

**Pull request merged**